### DISCUSSION TITLE
renovate pep621 bug since v38

### DISCUSSION THREAD

**Discussion by sigma67** [renovatebot/renovate#31162] [Request Help] [answered]
### What would you like help with?

I think I found a bug

### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us which platform (GitHub, GitLab, etc) and which version of Renovate.

Gitlab

### Please tell us more about your question or problem

Since v38 renovate does not correctly update pep621 dependencies anymore.

We suspect this is related to https://github.com/renovatebot/renovate/pull/28193/files

Minimal reproduction: https://github.com/sigma67/renovate-pep621-buildsystem-requires-reproduction

renovate is calling pdm incorrectly, it is skipping the group argument when using -dG (I picked one instance from the log and pasted it below).

```
 DEBUG: Executing command (repository=sigma67/renovate-pep621-buildsystem-requires-reproduction, branch=test-renovate/major-all-dependencies)
       "command": "pdm update --no-sync --update-eager -dG sphinx "
DEBUG: rawExec err (repository=sigma67/renovate-pep621-buildsystem-requires-reproduction, branch=test-renovate/major-all-dependencies)
       "err": {
         "cmd": "/bin/sh -c pdm update --no-sync --update-eager -dG sphinx ",
         "stderr": "Ignoring non-existing groups: sphinx\nDEPRECATED: `cross_platform` st
```

### Logs (if relevant)

https://github.com/sigma67/renovate-pep621-buildsystem-requires-reproduction/actions/runs/10670054300/job/29573260508

**Comment by rarkins** [comment]
Hi, please do this:

1. Drop the use of the .js file, instead all relevant config should be in renovate.json
2. Reduce to minimal config to reproduce the problem. e.g. is grouping necessary to reproduce? is rangeStrategy=bump necessary?

**Reply by sigma67** [reply]
Done.

It occurs without grouping. As far as I can tell mainly the dev-dependencies pyproject.toml group is affected, which would make sense considering the suspected MR I linked above

**Reply by rarkins** [reply]
It's still not minimal

**Reply by sigma67** [reply]
I have also removed the bump strategy and some pyproject.toml entries, it still occurs

**Reply by rarkins** [reply]
It's still not minimal, nor is it a valid renovate.json

**Reply by sigma67** [reply]
I have removed more unrelated stuff from renovate.json..

**Reply by rarkins** [reply]
How is lock file maintenance related? You have left that in the config however not mentioned it here

**Reply by sigma67** [reply]
It is related because as far as I know renovate only calls pdm to update the lockfile

**Reply by rarkins** [reply]
If you remove that block then does the other PR stop having problems? I think you'll find that it's not necessary for reproduction

**Reply by sigma67** [reply]
I removed the lockfileMaintenance and as expected the problem disappears. Job runtime is also reduce to below 30s from ~3 minutes before.

Note that the focus is on this incorrect call to pdm itself:

``pdm update --no-sync --update-eager -dG sphinx``

results in the pdm error

``Ignoring non-existing groups: sphinx``

.. because pdm is called incorrectly. The correct call would be

``pdm update --no-sync --update-eager -dG dev sphinx``

While grouping is no necessary to reproduce, it exacerbates the problem. Instead of calling pdm update once per group as in v37, pdm update is now called *once per dependency*, which results in massive runtime increases. Previously we were seeing <30 minutes, now it's beyond 4 hours. With no semantic difference in the PR content created by renovate.

Does this clear it up? Could you kindly acknowledge the bug?

**Reply by rarkins** [reply]
Are these two separate issues?

1. Wrong pdm update call as you wrote above (suspected bug)
2. pdm update command being run once per upgrade instead of once per branch (performance)

Right now it also doesn't make sense to me why lockFileMaintenance settings would influence a non-lock file maintenance PR for sphinx. It additionally doesn't make sense to me why the lock file maintenance PR remained open in your repo after you changed settings.

@secustor does the dev group bug part make sense to you?

**Reply by secustor** [reply]
I'm looking into a patch currently as I had a moment.  
https://github.com/secustor/renovate-reproduction-31162

Though I run into some weird behavior that PDM returns non-zero exit codes, if it creates a new virtual environment.
This seems to be connected to the update commands as else all our own doc system upgrades would fail too.

**Reply by sigma67** [reply]
https://github.com/renovatebot/renovate/blob/aa4419aa0b0d1fb8edffbac8e5948e9aa0f97209/lib/modules/manager/pep621/processors/pdm.spec.ts

I also checked this and the spec seems wrong to me. The following statements work:

```
pdm update --no-sync --update-eager -d sphinx 
pdm update --no-sync --update-eager -G dev sphinx
pdm update --no-sync --update-eager -dG dev sphinx
```

This one (the one in the spec) doesn't and results in the "Ignoring non-existing groups: sphinx" warning:

```
pdm update --no-sync --update-eager -dG sphinx
```

^ This command instead triggers a full lockfile update, not one updating only sphinx.

**Reply by sigma67** [reply]
The tests also target pdm 2.5 while renovate installs latest by default, so might make sense to update that to obtain accurate tests:

https://github.com/renovatebot/renovate/blob/aa4419aa0b0d1fb8edffbac8e5948e9aa0f97209/lib/modules/manager/pep621/processors/pdm.spec.ts#L90

**Reply by secustor** [reply]
Indeed the test case is not longer valid, but not because of that but rather the manager does not return anymore depNames in the style of `group/packageName`

No need to upgrade this as the test do not run PDM at all. This simply to check the generated commands.

**Reply by secustor** [reply, marked as answer]
https://github.com/renovatebot/renovate/pull/31181#issuecomment-2327948911

**Reply by sigma67** [reply]
Can confirm that this also fixes the exploding runtime we were seeing. Back to a quick 15 minutes now. Thanks for the quick fix!

**Reply by rarkins** [reply]
Thanks too for your help identifying and problem and solution!
